### PR TITLE
McAfee ESM v2

### PR DESCRIPTION
**Pull request by esharf**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/20675
fixes: https://github.com/demisto/etc/issues/27456
fixes: https://github.com/demisto/etc/issues/27903

## Description
rewrite McAfee ESM.

## Does it break backward compatibility?
   - [ ] Yes
   - [x] No

## Must have
- [x] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by glicht** [comment]
Add fromversion 5.0.0

**Comment by kirbles19** [comment]
@esharf Done.

File: Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/whitelist.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# pylint: skip-file
+# flake8: noqa
+# type: ignore
+_is_status_code_valid  # unused function (Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.py:39)
```
Comment (glicht) [inline Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/whitelist.py:4]: Why do you need this function _is_status_code_valid if it is not used?

Comment (esharf) [reply Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/whitelist.py:4]: Override a function in CommonServerPython

File: Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/whitelist.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# pylint: skip-file
```
[the remark this answers is not in the document]
Comment (esharf) [reply Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/whitelist.py:1]: Vulture lint accept only this format for disabling errors

File: Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.py [hunk 3]
```diff
@@ -0,0 +1,811 @@
...
+        self.__cache: Dict = {
+            'users': [],
+            'org': [],
+            'status': []
+        }
+
+    def __del__(self):
+        self.__logout()
+
+    def _is_status_code_valid(self, *other):    # noqa
```
Comment (glicht) [inline Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.py:39]: Do this change and then you can get rid of the whitelist.py of vulture.
```suggestion
    def _is_status_code_valid(self, *_other):    # noqa
```

File: Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/whitelist.py (hunk 1 above)
Comment (glicht) [reply Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/whitelist.py:4]: See my comment about naming other -> _other. It should allow you to delete the whitelist.py file then.

File: Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.yml [hunk 4]
```diff
@@ -0,0 +1,738 @@
...
+  name: timezone
+  required: false
+  type: 0
+- defaultvalue: '5'
+  display: fetchLimitAlarms
+  hidden: false
+  name: fetchLimitAlarms
+  required: false
+  type: 0
+- defaultvalue: '5'
```
Comment (yaakovi) [inline Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.yml:72]:
```suggestion
- defaultvalue: '50'
```

File: Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.yml [hunk 5]
```diff
@@ -0,0 +1,738 @@
...
+  name: fetchTime
+  required: false
+  type: 0
+- defaultvalue: '0'
+  display: McAfee ESM Timezone in hours (e.g if ESM timezone is +0300 => then insert
+    3)
+  name: timezone
+  required: false
+  type: 0
+- defaultvalue: '5'
```
Comment (yaakovi) [inline Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.yml:66]:
```suggestion
- defaultvalue: '50'
```

File: Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.yml [hunk 6]
```diff
@@ -0,0 +1,738 @@
...
+    - contextPath: McAfeeESM.Case.Summary
+      description: The summary of the case.
+      type: string
+    - contextPath: McAfeeESM.Case.Status
+      description: The status of the case.
+      type: string
+    - contextPath: McAfeeESM.Case.OpenTime
+      description: The date and time when the case was opened.
+      type: date
+    - contextPath: McAfeeESM.Case.Severity
```
Comment (yaakovi) [inline Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.yml:274]: what about `McAfeeESM.Case.Assignee`, `McAfeeESM.Case.Organization`, `McAfeeESM.Case.EventList`, `McAfeeESM.Case.Notes`?
all appeared in the original integration

File: Packs/McAfee_ESM_v2/TestPlaybooks/playbook-McAfee_ESM_v2_-_Test.yml [hunk 7]
```diff
@@ -0,0 +1,827 @@
...
+            iscontext: true
+      - - operator: isEqualString
+          left:
+            value:
+              simple: McAfeeESM.Case.ID
+            iscontext: true
+          right:
+            value:
+              simple: McAfeeESM.Case.ID
+            iscontext: true
```
Comment (yaakovi) [inline Packs/McAfee_ESM_v2/TestPlaybooks/playbook-McAfee_ESM_v2_-_Test.yml:612]: duplicate
```suggestion
      - - operator: isEqualString
          left:
            value:
              simple: McAfeeESM.Case.ID
            iscontext: true
          right:
            value:
              simple: McAfeeESM.Case.ID
            iscontext: true
```

File: Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.py [hunk 8]
```diff
@@ -0,0 +1,811 @@
...
+        'esm-add-case-status': McAfeeESMClient.add_case_status,
+        'esm-delete-case-status': McAfeeESMClient.delete_case_status,
+        'esm-get-case-list': McAfeeESMClient.get_case_list,
+        'esm-get-user-list': McAfeeESMClient.get_user_list,
+        'esm-acknowledge-alarms': McAfeeESMClient.acknowledge_alarms,
+        'esm-unacknowledge-alarms': McAfeeESMClient.unacknowledge_alarms,
+        'esm-delete-alarms': McAfeeESMClient.delete_alarm,
+        'esm-get-alarm-event-details': McAfeeESMClient.get_alarm_event_details,
+        'esm-list-alarm-events': McAfeeESMClient.list_alarm_events,
+        'esm-search': McAfeeESMClient.complete_search
```
Comment (yaakovi) [inline Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.py:795]:
```suggestion
        'test-module': client.test_module,
        'esm-fetch-fields': client.fetch_fields,
        'esm-get-organization-list': client.get_organization_list,
        'esm-fetch-alarms': client.fetch_alarms,
        'esm-add-case': client.add_case,
        'esm-get-case-detail': client.get_case_detail,
        'esm-edit-case': client.edit_case,
        'esm-get-case-statuses': client.get_case_statuses,
        'esm-edit-case-status': client.edit_case_status,
        'esm-get-case-event-list': client.get_case_event_list,
        'esm-add-case-status': client.add_case_status,
        'esm-delete-case-status': client.delete_case_status,
        'esm-get-case-list': client.get_case_list,
        'esm-get-user-list': client.get_user_list,
        'esm-acknowledge-alarms': client.acknowledge_alarms,
        'esm-unacknowledge-alarms': client.unacknowledge_alarms,
        'esm-delete-alarms': client.delete_alarm,
        'esm-get-alarm-event-details': client.get_alarm_event_details,
        'esm-list-alarm-events': client.list_alarm_events,
        'esm-search': client.complete_search
```

File: Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.py [hunk 9]
```diff
@@ -0,0 +1,811 @@
...
+    }
+    try:
+        if command == 'fetch-incidents':
+            client.fetch_incidents(params)
+        elif command in commands:
+            human_readable, context_entry, raw_response = commands[command](client)
```
Comment (yaakovi) [inline Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.py:801]:
```suggestion
            human_readable, context_entry, raw_response = commands[command]()
```

File: Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.py [hunk 10]
```diff
@@ -0,0 +1,811 @@
...
+            middle = middle.replace('~', '|')
+            data[i] = f'| {middle} |'
+
+        start = f'Search results\n{line_1}\n{line_2}\n'
+        return start + '\n'.join(data)
+    else:
+        return ''
+
+
+def crate_incident(raw_incidents: List[Dict], alarms: bool) -> List[Dict[str, Dict]]:
```
Comment (yaakovi) [inline Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.py:746]:
```suggestion
def create_incident(raw_incidents: List[Dict], alarms: bool) -> List[Dict[str, Dict]]:
```

File: Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.py [hunk 11]
```diff
@@ -0,0 +1,811 @@
...
+    for incident in range(len(raw_incidents)):
```
Comment (yaakovi) [inline Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.py:747]: the loop looks like C...
```suggestion
    for incident in raw_incidents:
```

File: Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.py [hunk 12]
```diff
@@ -0,0 +1,811 @@
...
+        alarm_id = str(raw_incidents[incident].get('id'))
+        summary = str(raw_incidents[incident].get('summary'))
+        incident_type = 'alarm' if alarms else 'case'
+        raw_incidents[incident] = {
```
Comment (yaakovi) [inline Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.py:751]: don't override the original list

File: Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.py [hunk 13]
```diff
@@ -0,0 +1,811 @@
...
+            human_readable = tableToMarkdown(name='Alarm events', t=context_entry)
+
+        return human_readable, {f'{CONTEXT_INTEGRATION_NAME}'
+                                f'AlarmEvent(val.ID && val.ID == obj.ID)': context_entry}, result
+
+    def complete_search(self):
+        time_out = self.args.get('timeOut', 30)
+        interval = 10
+        if time_out < interval:
+            raise ValueError('Time out to short.')
```
Comment (yaakovi) [inline Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.py:451]: in this case, use the timeout as the interval so the polling will run twice.

File: Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.yml (hunk 6 above)
Comment (esharf) [reply Packs/McAfee_ESM_v2/Integrations/McAfee_ESM_v2/McAfee_ESM_v2.yml:269]: It’s not a part of the data in this API call.

**Comment by Itay4** [comment]
@yaakovi @esharf whats the status here?

**Review by yaakovi — APPROVED** [review]

File: Tests/conf.json [hunk 14]
```diff
@@ -3041,7 +3047,11 @@
         "Cherwell Example Scripts - test": "Issue 27107",
         "Cherwell - test": "Issue 26780",
         "Infinipoint-Test": "Issue 27453",
-        "663e1ec3-5590-4e23-8d99-d8a4e0c22a52": "Issue 27454"
+        "663e1ec3-5590-4e23-8d99-d8a4e0c22a52": "Issue 27454",
+        "Init_McAfee_ESM_v2_Test": "issue #26187",
+
+        "_comment": "~~~ DEPRECATED ~~~",
+        "Endpoint Enrichment - Generic v2 - Test": "DEPRECATED"
```
Comment (ShahafBenYakir) [inline Tests/conf.json:3054]: @esharf The change was introduced here

**Pull request merged**